samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManagerRestoreUtil.java

```diff
@@ -156,53 +158,90 @@ private static CompletableFuture<Map<TaskName, Checkpoint>> restoreAllTaskInstan
-            }
-          }
+  private static CompletableFuture<Checkpoint> retryFailedRestore(CompletableFuture<Boolean> shouldRetryRestoreFuture,
+      TaskName taskName, Map<TaskName, Checkpoint> taskCheckpoints, CheckpointManager checkpointManager,
+      TaskRestoreManager taskRestoreManager, Config config, Map<TaskName, TaskInstanceMetrics> taskInstanceMetrics,
+      ExecutorService executor, File loggedStoreBaseDirectory, JobContext jobContext, ContainerModel containerModel,
+      String factoryName, Map<TaskName, Map<String, Set<String>>> taskBackendFactoryToStoreNames) {
+    return shouldRetryRestoreFuture.thenCompose(shouldRetryRestore -> {
+      if (shouldRetryRestore) {
+        // Try to restore with getDeleted flag - return a new task checkpoint
+        return restoreDeletedSnapshot(taskName, taskCheckpoints, checkpointManager, taskRestoreManager,
+            config, taskInstanceMetrics, executor,
+            taskBackendFactoryToStoreNames.get(taskName).get(factoryName), loggedStoreBaseDirectory, jobContext,
+            containerModel);
+      }
+      // if shouldRetryRestore is false, do not retry restore. This means restore completed successfully in the first try
+      // Return the old task checkpoint as no new checkpoint was created
+      return CompletableFuture.completedFuture(taskCheckpoints.get(taskName));
+    });
+  }
 
-          // Stop all persistent stores after restoring. Certain persistent stores opened in BulkLoad mode are compacted
-          // on stop, so paralleling stop() also parallelizes their compaction (a time-intensive operation).
-          try {
-            taskRestoreManager.close();
-          } catch (Exception e) {
-            LOG.error("Error closing restore manager for task: {} after {} restore", taskName,
-                ex != null ? "unsuccessful" : "successful", e);
-            // ignore exception from close. container may still be able to continue processing/backups
-            // if restore manager close fails.
-          }
-          return null;
-        });
-        taskRestoreFutures.add(taskRestoreFuture);
-      });
+  private static CompletableFuture<Void> cleanUpResources(CompletableFuture<Checkpoint> taskCheckpointFuture,
+      long startTime, SamzaContainerMetrics samzaContainerMetrics, TaskName taskName, TaskRestoreManager taskRestoreManager,
+      Map<TaskName, Checkpoint> newTaskCheckpoints) {
+    return taskCheckpointFuture.handle((checkpoint, throwable) -> {
+      // exception or not, update the restore time and close TaskRestoreManager as there will be no more retries.
+      updateRestoreTime(startTime, samzaContainerMetrics, taskName);
+      closeTaskRestoreManager(taskRestoreManager, taskName.getTaskName());
+      if (throwable != null) {
+        throw new SamzaException("Exception from attempt to restore deleted snapshot", throwable);
+      }
+      if (checkpoint != null) {
+        newTaskCheckpoints.put(taskName, checkpoint);
```

Comment:
Minor: Consider completing the future with the checkpoint and merging at the end, instead of mutating passed in maps. Not a blocker.